### DISCUSSION TITLE
Error on overriding overloaded method from stub file

### DISCUSSION THREAD

**Discussion by flafflar** [microsoft/pyright#7208] [Q&A] [answered]
Let's say there is a stub file containing a class declaration with an overloaded method:
```python
from typing import overload

class Parent:
    @overload
    def overloaded(self, x: int) -> int: ...
    @overload
    def overloaded(self, y: str) -> str: ...
```

Another file contains a child class of the Parent class that overrides the overloaded method:
```python
from typing import overload
from parent import Parent

class Child(Parent):
    @overload
    def overloaded(self, x: int) -> int: ...
    @overload
    def overloaded(self, y: str) -> str: ...
    def overloaded(self, x: int|str) -> int|str:
        return x
```
which gives the following error:
```
error: Overloaded implementation is not consistent with signature of overload 2
    Type "(self: Self@Child, x: int | str) -> (int | str)" cannot be assigned to type "(self: Self@Child, y: str) -> str"
      Parameter name mismatch: "y" versus "x" (reportInconsistentOverload)
```

Changing the parameter names to be the same across all overloads gives the following error:
```python
from typing import overload
from parent import Parent

class Child(Parent):
    @overload
    def overloaded(self, x: int) -> int: ...
    @overload
    def overloaded(self, x: str) -> str: ...
    def overloaded(self, x: int|str) -> int|str:
        return x
```
```
error: Method "overloaded" overrides class "Parent" in an incompatible manner
    Override does not handle all overloads of base method (reportIncompatibleMethodOverride)
```

This makes it impossible to override the overloaded method without getting an error. Is this a bug or is there some other way to correctly override the method?

**Comment by erictraut** [comment, marked as answer]
Is the stub correct in defining two overload signatures that have different parameter names? That looks like it's probably a bug in the stub. If so, it should be fixed there.

If this isn't a bug in the stub, then the overload definition indicates that the function accepts either a positional parameter of type `int | str`, a keyword parameter named `x` of type `int`, or a keyword parameter named `y` of type `str`. To accept that full range of possibilities, your implementation would need to accept `*args` and `**kwargs` and do a bunch of validation. Something like:

```python
    def overloaded(self, *args: int | str, **kwargs: int | str) -> int | str:
        if len(args) + len(kwargs) > 1:
            raise ValueError("Too many arguments")
        if len(args) == 1:
            arg = args[0]
        else:
            if "x" in kwargs:
                arg = kwargs["x"]
                if not isinstance(arg, int):
                    raise ValueError("x is wrong type")
            elif "y" in kwargs:
                arg = kwargs["y"]
                if not isinstance(arg, str):
                    raise ValueError("y is wrong type")
            else:
                raise ValueError("Unknown parameter name")

        return arg
```

**Reply by flafflar** [reply]
The different parameter names is not a bug, since it is allowed by the Python specification (see the [relevant PEP](https://peps.python.org/pep-0484/#function-method-overloading)). Shouldn't there be a more convenient way to overload that method instead of using args and kwargs?

**Reply by erictraut** [reply]
> Shouldn't there be a more convenient way to overload that method instead of using args and kwargs

This is a rare circumstance, so I think it would be difficult to justify amending the typing spec to provide a more convenient way to do this. If you're designing a new interface, I recommend against doing this. If you're retrofitting static types to a legacy interface, then you may need to jump through some hoops to make it work.

**Reply by flafflar** [reply]
Unfortunately, the interface that gave me this problem was PySide, so there is nothing I can do to change it. Thanks for your response anyways!
